File: reactive-tui-ts/src/examples/tui-demos/menu_demo.ts
#!/usr/bin/env bun

/**
 * Menu Widget Demo - TypeScript
 * 
 * Demonstrates the comprehensive menu widget functionality including:
 * - Basic menu navigation
 * - Hierarchical submenus
 * - Different item types (action, toggle, radio, separator, header)
 * - JSON configuration loading
 * - Keyboard navigation
 * - Theme integration
 */

import {
    menu, contextMenu, menuBar,
    menuItem, submenuItem, separatorItem, headerItem, toggleItem, radioItem,
    type MenuItem
} from '../../packages/tui-bun/src/widgets/menu';

console.log('🔥 Menu Widget Demo - TypeScript Implementation\n');

// Demo 1: Basic menu with different item types
console.log('📋 Demo 1: Basic Menu with Different Item Types');
console.log('=' .repeat(50));

const basicMenuItems: MenuItem[] = [
    headerItem({ id: 'header1', label: 'File Operations' }),
    menuItem({ id: 'new', label: 'New File', action: 'file.new', icon: '📄', shortcut: 'Ctrl+N' }),
    menuItem({ id: 'open', label: 'Open File', action: 'file.open', icon: '📂', shortcut: 'Ctrl+O' }),
    menuItem({ id: 'save', label: 'Save File', action: 'file.save', icon: '💾', shortcut: 'Ctrl+S' }),
    separatorItem('sep1'),
    headerItem({ id: 'header2', label: 'View Options' }),
    toggleItem({ id: 'lineNumbers', label: 'Show Line Numbers', state: true, shortcut: 'F11' }),
    toggleItem({ id: 'wordWrap', label: 'Word Wrap', state: false, shortcut: 'Alt+W' }),
    separatorItem('sep2'),
    headerItem({ id: 'header3', label: 'Theme Selection' }),
    radioItem({ id: 'dark', label: 'Dark Theme', group: 'theme', selected: true }),
    radioItem({ id: 'light', label: 'Light Theme', group: 'theme', selected: false }),
    radioItem({ id: 'terminal', label: 'Terminal Theme', group: 'theme', selected: false }),
];

const basicMenu = menu({
    id: 'basic-menu',
    items: basicMenuItems,
    orientation: 'vertical',
    colorTheme: 'dark',
    onItemSelect: (item) => {
        console.log(`✓ Selected: ${item.label} (${item.id})`);
        if (item.itemType.action) {
            console.log(`  Action: ${item.itemType.action}`);
        }
    }
});

console.log(`Menu created with ${basicMenuItems.length} items`);
console.log(`Menu ID: ${basicMenu.id}`);
console.log(`Focusable: ${basicMenu.focusable}`);

// Demo 2: Hierarchical submenu structure
console.log('\n🌳 Demo 2: Hierarchical Submenu Structure');
console.log('=' .repeat(50));

const editSubmenu: MenuItem[] = [
    menuItem({ id: 'undo', label: 'Undo', action: 'edit.undo', icon: '↶', shortcut: 'Ctrl+Z' }),
    menuItem({ id: 'redo', label: 'Redo', action: 'edit.redo', icon: '↷', shortcut: 'Ctrl+Y' }),
    separatorItem('edit-sep1'),
    menuItem({ id: 'cut', label: 'Cut', action: 'edit.cut', icon: '✂', shortcut: 'Ctrl+X' }),
    menuItem({ id: 'copy', label: 'Copy', action: 'edit.copy', icon: '📋', shortcut: 'Ctrl+C' }),
    menuItem({ id: 'paste', label: 'Paste', action: 'edit.paste', icon: '📌', shortcut: 'Ctrl+V' }),
];

const formatSubmenu: MenuItem[] = [
    headerItem({ id: 'format-header', label: 'Text Formatting' }),
    toggleItem({ id: 'bold', label: 'Bold', state: false, shortcut: 'Ctrl+B' }),
    toggleItem({ id: 'italic', label: 'Italic', state: false, shortcut: 'Ctrl+I' }),
    toggleItem({ id: 'underline', label: 'Underline', state: false, shortcut: 'Ctrl+U' }),
    separatorItem('format-sep1'),
    submenuItem({
        id: 'align',
        label: 'Text Alignment',
        icon: '⬌',
        items: [
            radioItem({ id: 'align-left', label: 'Left', group: 'alignment', selected: true }),
            radioItem({ id: 'align-center', label: 'Center', group: 'alignment', selected: false }),
            radioItem({ id: 'align-right', label: 'Right', group: 'alignment', selected: false }),
            radioItem({ id: 'align-justify', label: 'Justify', group: 'alignment', selected: false }),
        ]
    })
];

const hierarchicalMenuItems: MenuItem[] = [
    menuItem({ id: 'new-doc', label: 'New Document', action: 'file.new', icon: '📄', shortcut: 'Ctrl+N' }),
    submenuItem({ id: 'edit', label: 'Edit', icon: '✏', items: editSubmenu }),
    submenuItem({ id: 'format', label: 'Format', icon: '🎨', items: formatSubmenu }),
    separatorItem('main-sep1'),
    menuItem({ id: 'settings', label: 'Settings', action: 'app.settings', icon: '⚙', shortcut: 'Ctrl+,' }),
    menuItem({ id: 'about', label: 'About', action: 'app.about', icon: 'ℹ' }),
];

const _hierarchicalMenu = menu({
    id: 'hierarchical-menu',
    items: hierarchicalMenuItems,
    orientation: 'vertical',
    colorTheme: 'light',
    onItemSelect: (item) => {
        console.log(`✓ Selected: ${item.label} (${item.id})`);
    },
    onSubmenuOpen: (item) => {
        console.log(`📂 Opened submenu: ${item.label}`);
    },
    onSubmenuClose: (item) => {
        console.log(`📁 Closed submenu: ${item.label}`);
    }
});

console.log(`Hierarchical menu created with ${hierarchicalMenuItems.length} top-level items`);
console.log('Submenu structure:');
hierarchicalMenuItems.forEach(item => {
    if (item.itemType.type === 'submenu' && item.itemType.items) {
        console.log(`  └─ ${item.label}: ${item.itemType.items.length} items`);
        item.itemType.items.forEach(subItem => {
            if (subItem.itemType.type === 'submenu' && subItem.itemType.items) {
                console.log(`      └─ ${subItem.label}: ${subItem.itemType.items.length} items`);
            }
        });
    }
});

// Demo 3: JSON Configuration Loading
console.log('\n📄 Demo 3: JSON Configuration Loading');
console.log('=' .repeat(50));

const jsonMenuConfig = JSON.stringify([
    {
        "id": "file",
        "label": "File Menu",
        "itemType": { "type": "header" }
    },
    {
        "id": "new-project",
        "label": "New Project",
        "itemType": { "type": "action", "action": "project.new" },
        "icon": "🚀",
        "shortcut": "Ctrl+Shift+N"
    },
    {
        "id": "open-project",
        "label": "Open Project",
        "itemType": { "type": "action", "action": "project.open" },
        "icon": "📁",
        "shortcut": "Ctrl+Shift+O"
    },
    {
        "id": "recent",
        "label": "Recent Projects",
        "itemType": {
            "type": "submenu",
            "items": [
                {
                    "id": "project1",
                    "label": "AI CLI Project",
                    "itemType": { "type": "action", "action": "project.open.ai-cli" },
                    "icon": "🤖"
                },
                {
                    "id": "project2",
                    "label": "TUI Framework",
                    "itemType": { "type": "action", "action": "project.open.tui" },
                    "icon": "🖥"
                }
            ]
        },
        "icon": "📚"
    },
    {
        "id": "separator1",
        "label": "",
        "itemType": { "type": "separator" }
    },
    {
        "id": "auto-save",
        "label": "Auto Save",
        "itemType": { "type": "toggle", "state": true },
        "icon": "💾"
    }
], null, 2);

console.log('JSON Configuration:');
console.log(jsonMenuConfig);

const _jsonMenu = menu({
    id: 'json-menu',
    configData: jsonMenuConfig,
    configFormat: 'json',
    orientation: 'vertical',
    colorTheme: 'terminal',
    onItemSelect: (item) => {
        console.log(`✓ JSON Menu - Selected: ${item.label}`);
    }
});

console.log('\n✅ JSON menu loaded successfully!');

// Demo 4: Menu Bar (Horizontal)
console.log('\n🧭 Demo 4: Horizontal Menu Bar');
console.log('=' .repeat(50));

const menuBarItems: MenuItem[] = [
    submenuItem({
        id: 'file-menu',
        label: 'File',
        items: [
            menuItem({ id: 'new', label: 'New', shortcut: 'Ctrl+N' }),
            menuItem({ id: 'open', label: 'Open', shortcut: 'Ctrl+O' }),
            menuItem({ id: 'save', label: 'Save', shortcut: 'Ctrl+S' }),
        ]
    }),
    submenuItem({
        id: 'edit-menu',
        label: 'Edit',
        items: editSubmenu
    }),
    submenuItem({
        id: 'view-menu',
        label: 'View',
        items: [
            toggleItem({ id: 'sidebar', label: 'Show Sidebar', state: true }),
            toggleItem({ id: 'minimap', label: 'Show Minimap', state: false }),
            separatorItem('view-sep'),
            menuItem({ id: 'fullscreen', label: 'Toggle Fullscreen', shortcut: 'F11' }),
        ]
    }),
    submenuItem({
        id: 'help-menu',
        label: 'Help',
        items: [
            menuItem({ id: 'docs', label: 'Documentation', icon: '📖' }),
            menuItem({ id: 'shortcuts', label: 'Keyboard Shortcuts', icon: '⌨' }),
            separatorItem('help-sep'),
            menuItem({ id: 'about', label: 'About', icon: 'ℹ' }),
        ]
    }),
];

const _applicationMenuBar = menuBar({
    id: 'app-menu-bar',
    items: menuBarItems,
    colorTheme: 'dark',
    style: {
        showBorders: true,
        padding: 2
    },
    onItemSelect: (item) => {
        console.log(`🧭 Menu Bar - Selected: ${item.label}`);
    }
});

console.log(`Menu bar created with ${menuBarItems.length} top-level menus`);
console.log('Menu structure:');
menuBarItems.forEach(menu => {
    if (menu.itemType.type === 'submenu' && menu.itemType.items) {
        console.log(`  ${menu.label}: ${menu.itemType.items.length} items`);
    }
});

// Demo 5: Context Menu
console.log('\n🖱 Demo 5: Context Menu');
console.log('=' .repeat(50));

const contextMenuItems: MenuItem[] = [
    menuItem({ id: 'cut', label: 'Cut', action: 'edit.cut', icon: '✂', shortcut: 'Ctrl+X' }),
    menuItem({ id: 'copy', label: 'Copy', action: 'edit.copy', icon: '📋', shortcut: 'Ctrl+C' }),
    menuItem({ id: 'paste', label: 'Paste', action: 'edit.paste', icon: '📌', shortcut: 'Ctrl+V' }),
    separatorItem('context-sep1'),
    submenuItem({
        id: 'select',
        label: 'Select',
        icon: '🔍',
        items: [
            menuItem({ id: 'select-all', label: 'Select All', shortcut: 'Ctrl+A' }),
            menuItem({ id: 'select-word', label: 'Select Word', shortcut: 'Ctrl+D' }),
            menuItem({ id: 'select-line', label: 'Select Line', shortcut: 'Ctrl+L' }),
        ]
    }),
    separatorItem('context-sep2'),
    menuItem({ id: 'inspect', label: 'Inspect Element', action: 'dev.inspect', icon: '🔍' }),
];

const _rightClickMenu = contextMenu({
    id: 'context-menu',
    items: contextMenuItems,
    colorTheme: 'light',
    onItemSelect: (item) => {
        console.log(`🖱 Context Menu - Selected: ${item.label}`);
    }
});

console.log(`Context menu created with ${contextMenuItems.length} items`);

// Demo 6: Menu API Testing
console.log('\n🔧 Demo 6: Menu API Testing');
console.log('=' .repeat(50));

const dynamicMenu = menu({
    id: 'dynamic-menu',
    items: [
        headerItem({ id: 'dynamic-header', label: 'Dynamic Menu Demo' }),
        menuItem({ id: 'item1', label: 'Initial Item 1' }),
        menuItem({ id: 'item2', label: 'Initial Item 2' }),
    ],
    onItemSelect: (item) => {
        console.log(`🔧 Dynamic Menu - Selected: ${item.label}`);
    }
});

console.log('Initial menu state:');
console.log(`Items: ${dynamicMenu.api.getState().currentItems.length}`);

// Test adding items
console.log('\n📝 Testing API: Adding items...');
dynamicMenu.api.addItem(separatorItem('dynamic-sep'));
dynamicMenu.api.addItem(menuItem({ 
    id: 'added-item', 
    label: 'Dynamically Added Item', 
    action: 'dynamic.added',
    icon: '➕'
}));

console.log(`Items after adding: ${dynamicMenu.api.getState().currentItems.length}`);

// Test navigation
console.log('\n🧭 Testing API: Navigation...');
console.log(`Current selection: ${dynamicMenu.api.getState().selectedIndex}`);

dynamicMenu.api.nextItem();
console.log(`After next(): ${dynamicMenu.api.getState().selectedIndex}`);

dynamicMenu.api.nextItem();
console.log(`After next(): ${dynamicMenu.api.getState().selectedIndex}`);

dynamicMenu.api.previousItem();
console.log(`After previous(): ${dynamicMenu.api.getState().selectedIndex}`);

// Test JSON loading
console.log('\n📄 Testing API: JSON loading...');
const newJsonConfig = JSON.stringify([
    { "id": "loaded1", "label": "Loaded Item 1", "itemType": { "type": "action" } },
    { "id": "loaded2", "label": "Loaded Item 2", "itemType": { "type": "action" } },
    { "id": "sep", "label": "", "itemType": { "type": "separator" } },
    { "id": "toggle", "label": "Loaded Toggle", "itemType": { "type": "toggle", "state": true } }
]);

const loadResult = dynamicMenu.api.loadFromJson(newJsonConfig);
console.log(`JSON load result: ${loadResult}`);
console.log(`Items after JSON load: ${dynamicMenu.api.getState().currentItems.length}`);

// Test toggle and radio states
console.log('\n🔘 Testing API: Toggle and Radio states...');
dynamicMenu.api.setToggleState('toggle', false);
dynamicMenu.api.setRadioSelection('test-group', 'option1');

const finalState = dynamicMenu.api.getState();
console.log(`Toggle states:`, finalState.toggleStates);
console.log(`Radio selections:`, finalState.radioSelections);

console.log('\n🎉 All Menu Widget Demos Completed Successfully!');
console.log('=' .repeat(60));
console.log('✓ Basic menu with different item types');
console.log('✓ Hierarchical submenu structure');
console.log('✓ JSON configuration loading');
console.log('✓ Horizontal menu bar');
console.log('✓ Context menu');
console.log('✓ Dynamic menu API testing');
console.log('\n🚀 TypeScript Menu Widget is fully functional with feature parity to Rust implementation!');

File: examples/menu_demo.rs
use reactive_tui::prelude::*;
use reactive_tui::themes::colors::dark_theme;
use reactive_tui::widgets::*;

fn main() -> Result<()> {
  // Get terminal size dynamically
  let (term_width, _term_height) = crossterm::terminal::size().unwrap_or((400, 200));

  println!("🍽️ Menu Widget Demo\n");

  let layout = LayoutRect {
    x: 0,
    y: 0,
    width: term_width,
    height: 15,
  };
  let theme = dark_theme();

  // 1. Simple Menu with Builder API
  println!("📋 Simple Menu (Builder API):");
  let simple_menu = Menu::builder("simple-menu")
    .action("new", "New File", "file:new")
    .action("open", "Open File", "file:open")
    .action("save", "Save File", "file:save")
    .separator("sep1")
    .action("exit", "Exit", "app:exit")
    .build();

  println!("{}\n", simple_menu.render(&layout, Some(&theme)));

  // 2. Menu with Icons and Shortcuts
  println!("⚡ Menu with Icons and Shortcuts:");
  let icon_menu = Menu::builder("icon-menu")
    .item(MenuItem {
      id: "new".to_string(),
      label: "New Document".to_string(),
      item_type: MenuItemType::Action {
        action: "doc:new".to_string(),
      },
      icon: Some('📄'),
      shortcut: Some("Ctrl+N".to_string()),
      enabled: true,
      visible: true,
      css_classes: vec!["menu-item".to_string()],
      tooltip: Some("Create a new document".to_string()),
      data: std::collections::HashMap::new(),
    })
    .item(MenuItem {
      id: "open".to_string(),
      label: "Open Document".to_string(),
      item_type: MenuItemType::Action {
        action: "doc:open".to_string(),
      },
      icon: Some('📂'),
      shortcut: Some("Ctrl+O".to_string()),
      enabled: true,
      visible: true,
      css_classes: vec!["menu-item".to_string()],
      tooltip: Some("Open an existing document".to_string()),
      data: std::collections::HashMap::new(),
    })
    .item(MenuItem {
      id: "save".to_string(),
      label: "Save Document".to_string(),
      item_type: MenuItemType::Action {
        action: "doc:save".to_string(),
      },
      icon: Some('💾'),
      shortcut: Some("Ctrl+S".to_string()),
      enabled: true,
      visible: true,
      css_classes: vec!["menu-item".to_string()],
      tooltip: Some("Save the current document".to_string()),
      data: std::collections::HashMap::new(),
    })
    .build();

  println!("{}\n", icon_menu.render(&layout, Some(&theme)));

  // 3. Menu with Toggle and Radio Items
  println!("🔘 Menu with Toggle and Radio Items:");
  let interactive_menu = Menu::builder("interactive-menu")
    .header("view-header", "View Options")
    .toggle("word-wrap", "Word Wrap", false)
    .toggle("line-numbers", "Line Numbers", true)
    .separator("sep1")
    .header("theme-header", "Theme")
    .item(MenuItem {
      id: "theme-light".to_string(),
      label: "Light Theme".to_string(),
      item_type: MenuItemType::Radio {
        group: "theme".to_string(),
        selected: false,
      },
      icon: Some('☀'),
      shortcut: None,
      enabled: true,
      visible: true,
      css_classes: Vec::new(),
      tooltip: None,
      data: std::collections::HashMap::new(),
    })
    .item(MenuItem {
      id: "theme-dark".to_string(),
      label: "Dark Theme".to_string(),
      item_type: MenuItemType::Radio {
        group: "theme".to_string(),
        selected: true,
      },
      icon: Some('🌙'),
      shortcut: None,
      enabled: true,
      visible: true,
      css_classes: Vec::new(),
      tooltip: None,
      data: std::collections::HashMap::new(),
    })
    .build();

  println!("{}\n", interactive_menu.render(&layout, Some(&theme)));

  // 4. Hierarchical Menu with Submenus
  println!("🗂️ Hierarchical Menu with Submenus:");
  let file_submenu = vec![
    MenuItem {
      id: "new-text".to_string(),
      label: "Text File".to_string(),
      item_type: MenuItemType::Action {
        action: "file:new:text".to_string(),
      },
      icon: Some('📝'),
      shortcut: None,
      enabled: true,
      visible: true,
      css_classes: Vec::new(),
      tooltip: None,
      data: std::collections::HashMap::new(),
    },
    MenuItem {
      id: "new-markdown".to_string(),
      label: "Markdown File".to_string(),
      item_type: MenuItemType::Action {
        action: "file:new:markdown".to_string(),
      },
      icon: Some('📋'),
      shortcut: None,
      enabled: true,
      visible: true,
      css_classes: Vec::new(),
      tooltip: None,
      data: std::collections::HashMap::new(),
    },
  ];

  let hierarchical_menu = Menu::builder("hierarchical-menu")
    .submenu("file", "File", file_submenu)
    .action("edit", "Edit", "edit:menu")
    .action("view", "View", "view:menu")
    .action("help", "Help", "help:menu")
    .build();

  println!("{}\n", hierarchical_menu.render(&layout, Some(&theme)));

  // 5. Menu from JSON Configuration
  println!("📄 Menu from JSON Configuration:");
  let json_config = r#"[
        {
            "id": "git-status",
            "label": "Git Status",
            "type": "Action",
            "action": "git:status",
            "shortcut": "Ctrl+G S"
        },
        {
            "id": "git-commit",
            "label": "Git Commit",
            "type": "Action", 
            "action": "git:commit",
            "shortcut": "Ctrl+G C"
        },
        {
            "id": "git-push",
            "label": "Git Push",
            "type": "Action",
            "action": "git:push", 
            "shortcut": "Ctrl+G P"
        },
        {
            "id": "sep",
            "label": "",
            "type": "Separator"
        },
        {
            "id": "git-branch",
            "label": "Branch Management",
            "type": "Submenu",
            "items": [
                {
                    "id": "new-branch",
                    "label": "New Branch",
                    "type": "Action",
                    "action": "git:branch:new"
                },
                {
                    "id": "switch-branch", 
                    "label": "Switch Branch",
                    "type": "Action",
                    "action": "git:branch:switch"
                }
            ]
        }
    ]"#;

  let json_menu = Menu::from_json("json-menu", json_config)?;
  println!("{}\n", json_menu.render(&layout, Some(&theme)));

  // 6. Menu from YAML Configuration
  println!("📝 Menu from YAML Configuration:");
  let yaml_config = r#"
- id: debug-start
  label: Start Debugging
  type: Action
  action: debug:start
  shortcut: F5

- id: debug-step
  label: Step Over
  type: Action
  action: debug:step
  shortcut: F10

- id: debug-into
  label: Step Into
  type: Action
  action: debug:into
  shortcut: F11

- id: debug-out
  label: Step Out
  type: Action
  action: debug:out
  shortcut: Shift+F11

- id: sep
  label: ""
  type: Separator

- id: debug-breakpoints
  label: Breakpoints
  type: Submenu
  items:
    - id: toggle-breakpoint
      label: Toggle Breakpoint
      type: Action
      action: debug:breakpoint:toggle
      shortcut: F9
    
    - id: clear-breakpoints
      label: Clear All Breakpoints
      type: Action
      action: debug:breakpoint:clear
"#;

  let yaml_menu = Menu::from_yaml("yaml-menu", yaml_config)?;
  println!("{}\n", yaml_menu.render(&layout, Some(&theme)));

  // 7. Navigation Demo
  println!("🧭 Menu Navigation Demo:");
  let mut nav_menu = Menu::builder("nav-menu")
    .action("item1", "First Item", "action1")
    .action("item2", "Second Item", "action2")
    .action("item3", "Third Item", "action3")
    .build();

  println!("Initial selection (item 1):");
  println!("{}", nav_menu.render(&layout, Some(&theme)));

  // Navigate and show selection changes
  nav_menu.next_item();
  println!("After next_item() (item 2):");
  println!("{}", nav_menu.render(&layout, Some(&theme)));

  nav_menu.next_item();
  println!("After next_item() (item 3):");
  println!("{}", nav_menu.render(&layout, Some(&theme)));

  // Test activation
  if let Some(action) = nav_menu.activate_selected() {
    println!("Activated action: {action}\n");
  }

  // 8. Different Menu Styles
  println!("🎨 Menu Style Variations:");

  // Context menu style
  let context_menu = context_menu("context")
    .action("cut", "Cut", "edit:cut")
    .action("copy", "Copy", "edit:copy")
    .action("paste", "Paste", "edit:paste")
    .build();

  println!("Context Menu:");
  println!("{}", context_menu.render(&layout, Some(&theme)));

  // Menu bar style
  let menu_bar = menu_bar("menubar")
    .action("file", "File", "file:menu")
    .action("edit", "Edit", "edit:menu")
    .action("view", "View", "view:menu")
    .action("help", "Help", "help:menu")
    .build();

  println!("Menu Bar:");
  println!("{}", menu_bar.render(&layout, Some(&theme)));

  // Dropdown menu style
  let dropdown = dropdown_menu("dropdown")
    .action("option1", "Option 1", "option1")
    .action("option2", "Option 2", "option2")
    .action("option3", "Option 3", "option3")
    .build();

  println!("Dropdown Menu:");
  println!("{}", dropdown.render(&layout, Some(&theme)));

  println!("\n🎨 Menu Widget Demo Complete - All menu types and configurations demonstrated");

  Ok(())
}
